homestar-wasm/README.md

```diff
@@ -23,15 +23,1054 @@
+      {
+        "args": [[47, null]]
+      }
+      ```
+
+      `[47, null]` is converted into an `Ipld::List`, which is then translated
+      and passed into `fn` as an `Ok` case of the `result` argument with a
+      payload of `47` matching the `s32` type on the left.
+
+  * **WIT to IPLD Translation**:
+
+    Conversely, when a `result` value is returned from a WIT function, it can
+    be translated back into an `Ipld::List` of this specific structure.
+
+- `Err` (with a payload)
+
+  * **IPLD to WIT Translation**
+
+    **Example**:
+
+    ```wit
+    export fn: func(a: result<s32, string>) -> result<s32, string>;
+    ```
+
+    Given an JSON input for this function:
+
+      ```json
+      {
+        "args": [[null, "error message"]]
+      }
+      ```
+
+      `[null, "error message"]` is converted into an `Ipld::List`, which is
+      then translated and passed into `fn` as an `Err` case of the `result`
+      argument with a payload of `"error message"` matching the `string` type
+      on the right.
+
+  * **WIT to IPLD Translation**:
+
+      Conversely, when a `result` value is returned from a WIT function, it can
+      be translated back into an `Ipld::List` of this specific structure.
+
+- `Ok` case (without a payload)
+
+  * **IPLD to WIT Translation**
+
+    **Example**:
+
+    ```wit
+    export fn: func(a: result<_, string>) -> result<_, string>;
+    ```
+
+    Given an JSON input for this function:
+
+      ```json
+      {
+        "args": [[47, null]]
+      }
+      ```
+
+      `[47, null]` is converted into an `Ipld::List`, which is then translated
+      and passed into `fn` as an `Ok` case of the `result` argument. The payload
+      is ignored as it's not needed (expressed in the type as `_` above), so
+      `47` is not used.
+
+  * **WIT to IPLD Translation**:
+
+    **Here, when this specific `Ok` case is returned from a WIT function, it can
+    be translated back into an `Ipld::List`, but one structured as
+    `[1, null]` internally, which signifies the `Ok` (not error) case, but
+    discards the payload.
+
+- `Err` case (without a payload)
+
+  * **IPLD to WIT Translation**
+
+    **Example**:
+
+    ```wit
+    export fn: func(a: result<s32, _>) -> result<s32, _>;
+    ```
+
+    Given an JSON input for this function:
+
+      ```json
+      {
+        "args": [[null, "error message"]]
+      }
+      ```
+
+      `[null, "error message"]` is converted into an `Ipld::List`, which is
+      then translated and passed into `fn` as an `Err` case of the `result`
+      argument. The payload is ignored as it's not needed (expressed in the type
+      as `_` above), so `"error message"` is not used.
+
+  * **WIT to IPLD Translation**:
+
+    **Here, when this specific `Err` case is returned from a WIT function, it
+    can be translated back into an `Ipld::List`, but one structured as
+    `[null, "error message"]` internally, which signifies the `Err` (error)
+    case, but discards the payload.
```

Comment:
```suggestion
    **Here, when this specific `Err` case is returned from a WIT function, it
    can be translated back into an `Ipld::List`, but one structured as
    `[null, 1]` internally, which signifies the `Err` (error) case, with
    the `1` payload discarded.**
```
Can't remember for sure, but I think we also use `1` as a placeholder payload here too.